### ISSUE TITLE
Implement simple http relay interface

### ISSUE THREAD

**Issue by phifty** [simia-tech/rust-delix#3]
The interface should open a port and listen for http requests. It's not clear yet how the request is translated to the delix network. More work on the node might be need in order to work on this issue.

**Comment by coopernurse** [comment]
I think we should write a short design doc of the messaging patterns we wish to support before we write the HTTP interface.

For example, do we want to support all the patterns from Iris (request/response, pub/sub, tunnel, broadcast)?

**Comment by phifty** [comment]
In the first version, I would only focus on request/response. So the HTTP interface would only handle this pattern.

If we implement pub/sub in the next version, I could imagine two options. Either, we extend the HTTP interface with websockets in the way that a client would subscribe to a topic by opening a WS connection. Or, we decide that the HTTP relay just don't support other patterns at all. I could life with that too.

We should indeed build an architecture that support running multiple relay module implementations in parallel ;-) Everyone could build then a setup that fits his needs.

Anyway, since the documentation of those concepts is important, I've started a Wiki page about messaging patterns.

**Comment by moea** [comment]
Assuming that the node exposes some HTTP resources for the  creation of subscriptions & the publishing of messages, it seems like pub/sub could probably also be implemented comfortably using vanilla HTTP - the application supplies an optional endpoint (e.g. `/topic/my-topic-name`) with each subscribe HTTP request, and messages are then delivered to the application as individual POST requests for the lifetime of the subscription.  Incoming messages could be adorned with some custom headers, to handle the case where multiple topics are routed to the same endpoint, or where a custom endpoint wasn't provided with the subscription request.

The custom endpoint stuff could be avoided if we felt comfortable enforcing a particular resource layout on applications - e.g. it's always `/topic/<topic-name>` which receives the POSTs.

**Comment by coopernurse** [comment]
That's an interesting idea and is similar to Amazon SNS.  If we decide to implement this, I wonder if we should also allow request/reply messages to proxy directly via HTTP.

This would allow developers to write HTTP based services as they normally do, and merely send 'registration' requests to Delix when the application starts (and ideally 'deregistration' requests when exiting).

I'm working on a markdown document that expands on these ideas.  I'll post it as a gist and link to it from this ticket and we can discuss it in more detail.

**Comment by phifty** [comment]
> If we decide to implement this, I wonder if we should also allow request/reply messages to proxy directly via HTTP.

Hmm... This is exactly what I had in mind with the HTTP relay :-) But I guess this means, the Wiki doc that I wrote about it is not really clear. Maybe you can help improving it. You should have write access to the Wiki.

**Comment by coopernurse** [comment]
Happy New Year.    I've written up two wiki pages that propose how we could use HTTP to relay request/reply messages.
- [Registration](https://github.com/simia-tech/rust-delix/wiki/HTTP-API--%7C-Registration) - Specifies how to register and deregister a service.  Also describes what a relayed message to the service would look like.
- [Send](https://github.com/simia-tech/rust-delix/wiki/HTTP-API-%7C-Send) - Specifies how to send a message to a service.

Currently only Request/Reply is specified, as I think pub/sub requires more discussion (particularly whether we want to support message groups, or offline store/forward, etc).

Let's discuss my spec here.  The main points of discussion include:
- How to specify the service name.  I chose to use a `X-Delix-Service` header, rather than using the `Host` header, just to avoid potential issues with people using `Host` for other purposes.
- API endpoint naming.  I chose `/api/v1/` as a prefix to give us the ability to version the API.  I know some people have strong opinions about whether you should have URLs like that.
- JSON request / response key names.

What do you guys think?

**Comment by moea** [comment]
Happy New Year!  This looks good.  At best semi-relevant high level stuff:

Does the use of subscription identifiers imply that we want to allow a single service instance to repeatedly register a single endpoint as a callback for the same service+operation (e.g. multiple request/reply registrations for service X at localhost:8000/x-handler)?  If that's true, it seems like the subscription identifier might also go in the request headers passed to the endpoint - and that a create-if-not-exists feature (which hey, HTTP) may one day come in handy for subscriptions.

If not, registration might also be an idempotent operation?  On the other side, I wonder if we could we could perhaps optionally use the service, endpoint and operation type as a means of identifying the subscription to tear down (so client code has one fewer thing to remember).

**Comment by coopernurse** [comment]
Good points.  For req/rep subscriptions, this string would be unique:

`reqprep:addr:port:path:service`

So we could use that internally to track the subscription, and omit the `regId` UUID completely.  This would mean that the register and deregister calls would have the same payload format.  That seems simple enough for the client, as there would be no state to manage.

**Comment by phifty** [comment]
Happy New Year to you too! I hope you had a couple of nice holidays.
- I'm in for the custom `X-Delix-Service` header. You're right, that `Host` might be used for something else.
- It might be good to modify the HTTP request as less as possible. So the `path` that goes in on one side is equal to the `path` that comes out. So either remove the `path` field completely or make it optional.
- In terms of naming, I vote always for non-shortcut version. Most use code completion nowadays and it will prevent any misunderstanding. But it's of course also a esthetical question ;-)

Apart from that, I had also something non-dynamic in mind for the first version. I though we might just define the services in the configuration file and not make them changeable at runtime. E.g.

```
[[relay]]
type = "http"
address = "localhost:4000"

[[relay.service]]
name = "session"
address = "localhost:8080"

[[relay.service]]
name = "user"
address = "localhost:8090"
```

My argument is, that those service usually also not dynamically change the routes they provide. The HTTP routes usually exists as long as the service is running. The only thing that might changes is that the service is running at all. But I would vote for the optimistic approach and just assume that the service is running and otherwise respond some `ServiceDown` error. This would be definitely good enough for all of my use cases.

As a bonus, there would be not need to provide any API for registration/deregistration at all (sorry for being such a stopper here). But maybe I miss out some cases here. Would this go along with your use cases?

**Comment by coopernurse** [comment]
Hi there,

`path` is optional on the register request.  It defaults to /   

I currently have some components where a single physical process exposes multiple logical services, each with a separate endpoint path, so being able to optionally specify the path would be useful.

For my use case, dynamic service registration will be important.  I have a bunch of services that I want to connect, but I don't want to explicitly manage which machines run which service.  When a service starts up, it would connect to the local Delix relay and register itself with its logical name.  When it shuts down (e.g. via SIGTERM) it would deregister.

One of my goals is to try to lower the amount of work required to introduce a new service into the system, and one of our current problems is managing all the configuration.

I'm not opposed to also allowing static configuration of the services.  When the app starts, we can seed the registry based on configuration, but allow that state to be mutated via the HTTP API.

Of course, this means that we'll have to communicate new service registrations across the Delix relay cluster.

If this sounds reasonable I can start a feature branch for this work.  Since I'm 100% new to Rust, I'll likely need some help with the implementation.

**Comment by phifty** [comment]
I agree that less configuration is a good thing. But I've the feeling with the dynamic solution, the configuration would be just moved into the service. I mean the service will at least have to know via which interface to use when sending the registration.

Anyway, I don't want to get in your way. And I definitely think we can do both. But I think, it would make sense to build two different relay modules since the dataset and the functionality will be quite different.

I could offer to implement the basic trait, configuration and loader structure and then work on the `http_static` (on a different issue) and you implement the `http_dynamic` relay. Sounds that good to you?

I've added you on jabber too and I'll try to help if you catch me there.

**Comment by coopernurse** [comment]
I need to read the existing code in more detail.  It seems that there's two separate concepts:
- Managing the service mapping (logical names to HTTP IP addresses/ports)
- Routing messages between clients and the service implementations

That static vs dynamic question affects the first item, but hopefully not the second.  That is, the actual act of routing messages should be separate from the code that manages the service mapping.  Does that sound right?

Dynamic configuration would just be a superset of the static implementation.  That is, a node would load its static configuration at startup (which may be empty), and then the HTTP API would allow that configuration to be modified at runtime.

If that makes sense, then I agree we could start with a static implementation, perhaps stubbing out the hooks that would allow the service mapping to be mutated (which would be used by the dynamic HTTP API later).

**Comment by phifty** [comment]
> That static vs dynamic question affects the first item, but hopefully not the second. That is, the actual act of routing messages should be separate from the code that manages the service mapping. Does that sound right?

That is correct. The message routing is done in the transport module and service mapping should be done in the relay implementation.

> Dynamic configuration would just be a superset of the static implementation. That is, a node would load its static configuration at startup (which may be empty), and then the HTTP API would allow that configuration to be modified at runtime.

Theoretical yes :-) I - as a user - would just get confused, if I configure some services in a config file, which I can delete than afterwards. Since they would not be deleted from the file, they would come back on a restart. I think it would be (let's name it) strait-forward to have two separate datasets. One static that can't be modified at runtime and one dynamic that is only managed via an API.

And since the static would not need an API at all, I thought separating the code as well in two modules might be a good idea.

I'll open another issue for the `http_static` module.

**Comment by coopernurse** [comment]
Ah, that makes sense.  

Where is the code that manages the routing table?  

What HTTP server lib should we use?  I see some suggestions to use Hyper:

https://github.com/hyperium/hyper

**Comment by phifty** [comment]
The routing happens mostly in the service_map.rs.

For the `http_static`, I've have currently two concurrent versions. One that uses hyper and one that pipes just tcp stream without doing much parsing. It only catches the `X-Delix-Service` header. I'm not sure which one will win in the end. Surprisingly the hyper-version has more code since the request-serialization is not build in and doesn't work yet. The raw-version is basically working, but doesn't support chunked transfers yet. Some of the code (not all) is on the `relay` branch.

**Comment by coopernurse** [comment]
Thanks for walking me through this.  I'll start looking at the `relay` branch.

**Comment by phifty** [comment]
I would close this issue and open a new one for the `http_dynamic`.
